feat(client): progress summary on the main courses page

arantebw opened the pull request:
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #41403

DONE:
- [x] Display the progress summary of the entire curriculum
- [x] Display an icon (check) if the certification was earned
- [x] Style the icon (check) appropriate to user theme
- [x] Display the progress summary for every _superblock_ page
- [x] Display the completed challenges of the current _superblock_
- [x] Display the complete projects of the current _superblock_
- [x] Implement a better UI for the progress summary

Please, advise if I also need to create an integration test for this PR.

Screenshots of the implemented UI:

![image](https://user-images.githubusercontent.com/8081121/216883204-e1ac7908-86cf-46aa-ad85-904840ce4309.png)

![image](https://user-images.githubusercontent.com/8081121/214790467-0dc35df5-1da2-4d4a-905e-d221ff4ca71c.png)

arantebw committed:
feat(components): add progress indicator component

arantebw committed:
feat(components): fetch certificate count on static query

arantebw committed:
feat(components): compute the certificates earned percentage

arantebw committed:
fix(component): import current cert object correctly

arantebw committed:
refactor(props): rename Props to style guide recommended format

arantebw committed:
fix(data): catch if data object is empty yet

arantebw committed:
refactor(component): move progress indicator component to learn page

arantebw committed:
refactor(component): conditionally render the progress indicator component

arantebw committed:
feat(component): add new certificate earned component

arantebw committed:
feat(component): display the check mark if cert was earned

arantebw committed:
feat(check): style the check mark for certs earned

arantebw committed:
chore(check): change icon size to 24px

arantebw committed:
refactor(component): remove unused styles.css file

arantebw committed:
refactor(progress-indicator): move styles.css to progress-indicator.css

arantebw committed:
refactor(styles): use meaningful class names for styles

arantebw committed:
feat(superblock): display the current superblock name in progress summary

arantebw committed:
feat(component): conditionally render overall and/or superblock progress

arantebw committed:
feat(component): get the completed challenges count of current superblock

arantebw committed:
feat(superblock): filter the number of projects completed per superblock

arantebw committed:
feat(progress): reuse the progress bar component

arantebw committed:
feat(styles): implement a nice ui for the progress summary section

bbsmooth commented:
This is looking very nice. But of course I do have some accessibility issues to bring up (because, hey, it's me).

I think we have duplicate information here for screen reader users. We use text above the progress bar to state how many challenges have been completed out of the total number of challenges and also the percentage completed. And then the progress bar basically just repeats that percentage. We should choose which one we want to use for screen readers: either the text percentage above the progress bar or the progress bar itself. I think I would probably vote for the text above the progress bar and then just hide the progress bar completely using `aria-hidden="true"`. 

And if we are going to hide the progress bar from AT then we don't really need to use an actual progress bar, we can just use a `div` inside of a `div` and set the width on the inner `div` to the percentage. You wouldn't even need to hide this from screen readers since there is no actual content. Just a thought.

bbsmooth commented:
One more issue. The white background on the bar does not have enough contrast with the light gray background. I think we will need to add a darker border around the entire bar.

bbsmooth commented:
A question. What is the rationale for using an `article` element to display the statistics?

bbsmooth commented:
Just a thought. What would you think about placing the progress summary above the Courses heading on the block pages? To me it seems like the progress summary is sort of its own thing and not a course that should be listed under the Courses heading.

arantebw committed:
refactor(progress-indicator): move location of the progress indicator component in superblock

arantebw commented:
> A question. What is the rationale for using an article element to display the statistics?
My thought process was to isolate portions of a `section` into two (2) child `article`s and semantically identify them though I did change these to child `section`s instead realizing it might really help in terms of accessibility.

arantebw commented:
> And if we are going to hide the progress bar from AT then we don't really need to use an actual progress bar, we can just use a div inside of a div and set the width on the inner div to the percentage. You wouldn't even need to hide this from screen readers since there is no actual content. Just a thought.

If I understand this correctly, I can remove the `ProgressBar` component and then create a pseudo-progress-bar consisting of a nested `<div>`s wherein the child `<div>`'s width depends on the completed items?

bbsmooth commented:
> I did change these to child sections instead realizing that the article might not really help in terms of accessibility.

That's fine. Section elements are harmless here. The issue with article elements is that some screen readers will call them out, thus adding to the verbosity. And I don't think this information really fit the definition of an article either, so that had the potential for further confusion.

We are good now.

bbsmooth commented:
> If I understand this correctly, I can remove the ProgressBar component and then create a pseudo-progress-bar consisting of a nested <div>s wherein the child <div>'s width depends on the completed items?

Exactly. But I'm not demanding this. There may be other reasons to use the ProgressBar component. I'll let others make that decision. I'm just suggesting that accessibility is not one of those reasons since we are just going to hide it from AT any way (assuming we go with my suggestion above).

Sboonny commented on client/src/components/CertificateEarned/index.tsx:
```diff
@@ -0,0 +1,117 @@
+      break;
+  }
+
+  return certsMap[superBlock] ? <CircleCheckRegular /> : <></>;
```
The label is ignored by screen readers, I think it can do more by adding `aria-label` with the text `earned`.

```suggestion
  return certsMap[superBlock] ? <CircleCheckRegular aria-label='earned' /> : <></>;
```

bbsmooth replied:
If you want to use `aria-label` on this svg then you'll need to add `role="img"` to the svg as well. Without the role, most screen readers will ignore the svg completely, even with `aria-label` on it.

Sboonny commented on client/src/components/ProgressIndicator/progress-indicator.css:
```diff
@@ -0,0 +1,45 @@
+.progress-summary {
+  border: 3px solid var(--secondary-color);
+  padding: 16px;
+  margin-bottom: 10px;
+  display: flex;
+  flex-direction: column;
+  row-gap: 16px;
+}
```
```suggestion
.progress-summary {
  border: 3px solid var(--secondary-color);
  padding: 16px;
  margin-bottom: 10px;
}
```

arantebw replied:
Done. 🙌🏽

Sboonny commented on client/src/components/ProgressIndicator/progress-indicator.css:
```diff
@@ -0,0 +1,45 @@
+  margin: 0;
+}
+
+.progress-summary__section {
+  display: flex;
+  flex-direction: column;
+  row-gap: 16px;
+}
```
```suggestion
.progress-summary,
.progress-summary__section {
  display: grid;
  row-gap: 16px;
}
```

arantebw replied:
Done. 🙌🏽

Sboonny commented on client/src/components/ProgressIndicator/progress-indicator.css:
```diff
@@ -0,0 +1,45 @@
+}
+
+.progress {
+  overflow: hidden;
```
```suggestion
```

Sboonny commented on client/src/components/ProgressIndicator/progress-indicator.css:
```diff
@@ -0,0 +1,45 @@
+  border-radius: 0;
+}
+
+.progress-bar {
+  background-color: var(--blue-mid);
+  float: left;
+  height: 100%;
+  font-size: 16px;
+  line-height: 25px;
+  color: #fff;
+  text-align: center;
+  box-shadow: inset 0 -1px 0 rgb(0 0 0 / 15%);
```
```suggestion
.progress-bar {
  background-color: var(--blue-mid);
  height: 100%;
  font-size: 16px;
  color: #fff;
  box-shadow: inset 0 -1px 0 rgb(0 0 0 / 15%);
```

arantebw replied:
Done. 🙌🏽

Sboonny commented on client/src/components/ProgressIndicator/index.tsx:
```diff
@@ -0,0 +1,282 @@
+          <section>
+            <div className='progress-summary__completed'>
+              <span>
+                {earnedCertificateCount}/{allCertificateCount} certificates
+                earned
+              </span>
+              <span>{completedCertificatePercentage}%</span>
+            </div>
+            <div>
+              <ProgressBar now={completedCertificatePercentage} />
+            </div>
+          </section>
+        </section>
+      )}
+      {superBlock !== '' && (
+        <section className='progress-summary__section'>
+          <section>
+            <div className='progress-summary__completed'>
+              <span>
+                {superBlockCompletedChallengesCount}/
+                {superBlockTotalChallengesCount} challenges completed
+              </span>
+              <span>{superBlockCompletedChallengesPercent}%</span>
+            </div>
+            <div>
+              <ProgressBar now={superBlockCompletedChallengesCount} />
+            </div>
+          </section>
+          <section>
+            <div className='progress-summary__completed'>
+              <span>
+                {superBlockCompletedProjectsCount}/
+                {superBlockTotalProjectsCount} projects completed
+              </span>
+              <span>{superBlockCompletedProjectsPercent}%</span>
+            </div>
+            <div>
+              <ProgressBar now={superBlockCompletedProjectsPercent} />
+            </div>
+          </section>
```
```suggestion
            </div>
            <ProgressBar now={completedChallengePercentage} />
          </section>
          <section>
            <div className='progress-summary__completed'>
              <span>
                {earnedCertificateCount}/{allCertificateCount} certificates
                earned
              </span>
              <span>{completedCertificatePercentage}%</span>
            </div>
            <ProgressBar now={completedCertificatePercentage} />
          </section>
        </section>
      )}
      {superBlock !== '' && (
        <section className='progress-summary__section'>
          <section>
            <div className='progress-summary__completed'>
              <span>
                {superBlockCompletedChallengesCount}/
                {superBlockTotalChallengesCount} challenges completed
              </span>
              <span>{superBlockCompletedChallengesPercent}%</span>
            </div>
            <ProgressBar now={superBlockCompletedChallengesCount} />
          </section>
          <section>
            <div className='progress-summary__completed'>
              <span>
                {superBlockCompletedProjectsCount}/
                {superBlockTotalProjectsCount} projects completed
              </span>
              <span>{superBlockCompletedProjectsPercent}%</span>
            </div>
            <ProgressBar now={superBlockCompletedProjectsPercent} />
          </section>
```

Sboonny reviewed (commented):
Other than the `aria-label` suggestion, the others are just clean up, nicely done. 👍

bbsmooth commented on client/src/components/ProgressIndicator/index.tsx:
```diff
@@ -0,0 +1,282 @@
+
+  return (
+    <div className='progress-summary'>
+      <h3 className='progress-summary__main-header'>
+        {headerText} Progress Summary
+      </h3>
```
```suggestion
      <h2 className='progress-summary__main-header'>
        {headerText} Progress Summary
      </h2>
```

Now that the progress summary is always above the certifications/courses, it needs to be an `h2` heading to fit into the hierarchy properly.

bbsmooth commented on client/src/components/ProgressIndicator/index.tsx:
```diff
@@ -0,0 +1,282 @@
+              <span>{completedChallengePercentage}%</span>
+            </div>
+            <div>
+              <ProgressBar now={completedChallengePercentage} />
```
```suggestion
              <ProgressBar
                aria-hidden='true'
                now={completedChallengePercentage}
              />
```

arantebw replied:
Done. 🙌🏽

bbsmooth commented on client/src/components/ProgressIndicator/index.tsx:
```diff
@@ -0,0 +1,282 @@
+              <span>{completedCertificatePercentage}%</span>
+            </div>
+            <div>
+              <ProgressBar now={completedCertificatePercentage} />
```
```suggestion
              <ProgressBar
                aria-hidden='true'
                now={completedChallengePercentage}
              />
```

bbsmooth commented on client/src/components/ProgressIndicator/index.tsx:
```diff
@@ -0,0 +1,282 @@
+              <span>{superBlockCompletedChallengesPercent}%</span>
+            </div>
+            <div>
+              <ProgressBar now={superBlockCompletedChallengesCount} />
```
```suggestion
              <ProgressBar
                aria-hidden='true'
                now={completedChallengePercentage}
              />
```

arantebw replied:
Done. 🙌🏽

bbsmooth commented on client/src/components/ProgressIndicator/index.tsx:
```diff
@@ -0,0 +1,282 @@
+              <span>{superBlockCompletedProjectsPercent}%</span>
+            </div>
+            <div>
+              <ProgressBar now={superBlockCompletedProjectsPercent} />
```
```suggestion
              <ProgressBar
                aria-hidden='true'
                now={completedChallengePercentage}
              />
```

bbsmooth commented on client/src/pages/learn.tsx:
```diff
@@ -105,6 +114,13 @@ function LearnPage({
+                username={username}
+                pathname={path}
+              />
+            )}
```
```suggestion
            )}
            <h2 className='sr-only'>{t('settings.headings.certs')}</h2>
```

Now that we have a heading for the progress summary, I think it makes sense to have one for the certifications. We can visually hide it so it doesn't ruin the aesthetics of the page.

arantebw replied:
Done. 🙌🏽

arantebw committed:
refactor(circlecheckregular): add aria label to the svg element

arantebw committed:
refactor(progress-indicator): make use of css grid instead of css flexbox

arantebw committed:
refactor(progress-indicator): remove the overflow property

arantebw committed:
refactor(progress-indicator): remove the float, ling-height, text-align props in .progress-bar class

arantebw committed:
Update client/src/components/ProgressIndicator/index.tsx

Co-authored-by: Muhammed Mustafa <muhammed@freecodecamp.org>

arantebw committed:
Update client/src/components/ProgressIndicator/index.tsx

Co-authored-by: Bruce Blaser <bbsmooth@gmail.com>

arantebw committed:
refactor(progress-indicator): apply the aria-hidden attribute to ProgressBar invocations

arantebw committed:
feat(learn.tsx): add an sr-only header for certifications

Sboonny commented on client/src/components/CertificateEarned/index.tsx:
```diff
@@ -0,0 +1,121 @@
+  }
+
+  return certsMap[superBlock] ? (
+    <CircleCheckRegular aria-label='earned' />
```
I don't know why there is an error

![image](https://user-images.githubusercontent.com/88248797/218007625-daafddcd-0cae-4517-b49d-190971bd3f3a.png)

I am happy to ignore it, because isn't part of the superblock so it doesn't make sense to me

---

Bruce's suggestion is missed, because of many suggestion 👍 

https://github.com/freeCodeCamp/freeCodeCamp/pull/49010#discussion_r1100722603

> If you want to use `aria-label` on this svg then you'll need to add `role="img"` to the svg as well. Without the role, most screen readers will ignore the svg completely, even with `aria-label` on it.



```suggestion
    <CircleCheckRegular role='img' aria-label='earned' />
```

arantebw replied:
I put it on the `<svg>` element. But yeah, I moved it here. 😅

Sboonny commented on client/src/components/CertificateEarned/index.tsx:
```diff
@@ -0,0 +1,121 @@
+  user: User;
+}
+
+const CertificateEarned = (props: CertificateEarnedProps): JSX.Element => {
+  const { superBlock, user } = props;
```
maybe the type is lost, but it still doesn't make sense, does changing it to this help with the type error?

```suggestion
const CertificateEarned = ({superBlock, user}: CertificateEarnedProps) => {
```

arantebw replied:
I tried to merge this but still fails the test. This is what is causing the linting to fail:
![image](https://user-images.githubusercontent.com/8081121/218233657-0757110f-0867-41b3-9b01-d1d8f0504536.png)

arantebw replied:
I don't think that the `SuperBlocks.TheOdinProject` still exists.

arantebw replied:
Will updating this PR branch with `main` resolve this issue? 🤔

Sboonny replied:
Let's test that, if it doesn't work, we can try turning `certsMap` into a Map.

arantebw committed:
refactor(circle-check-regular): add the role prop

Sboonny reviewed (approved):
LGTM 🎉, thank you for your contribution to the page! 👍.

arantebw committed:
Update client/src/components/CertificateEarned/index.tsx

Type error in SuperBlocks object.

Co-authored-by: Muhammed Mustafa <muhammed@freecodecamp.org>

arantebw committed:
chore: sync remote to local PR branch

Merge branch 'feat/progress-indicator' of github.com:arantebw/freeCodeCamp into feat/progress-indicator

arantebw committed:
style(certificate-earned): add some spacing on prop definitions

Commit:
Merge branch 'main' into feat/progress-indicator

Sboonny commented:
let me add `isOdinCert` real quick here https://github.com/arantebw/freeCodeCamp/blob/738fe0fa67073ef342f24ea627730a2d39d998f2/api-server/src/server/boot/certificate.js#L198-L215, to open the pr

arantebw committed:
Merge branch 'main' into feat/progress-indicator

Sboonny commented:
@arantebw, sorry for the delay, we have discussed this issue and we came to the conclusion that we need to do additional mock up for the progress card, and some refactoring to make it easier to make this changes, thank you for your patience

arantebw commented:
> @arantebw, sorry for the delay, we have discussed this issue and we came to the conclusion that we need to do additional mock up for the progress card, and some refactoring to make it easier to make this changes, thank you for your patience

No worries. 👍🏼 Please let me know if there's anything I can help you with.

ahmaxed committed:
Merge branch 'main' into feat/progress-indicator

arantebw committed:
Merge branch 'main' into feat/progress-indicator

arantebw commented:
My PR is still failing several integration tests. Is there anything that I need to do on my end?

Sboonny commented:
> My PR is still failing several integration tests. Is there anything that I need to do on my end?

Nope, the PR is still using the flaky Firefox tests, a merge/rebase the main to the branch should sort, don't worry about it.

raisedadead committed:
Merge branch 'main' into feat/progress-indicator

raisedadead commented:
The tests need to be updated on this PR. The E2E tests need to be fixed (they are not flaky).

@Sboonny and @ahmadabdolsaheb, please: 

A) Prioritise the mockup and other logistics the OP needs to continue their work

OR

B) Please merge these changes with a feature flag, and bring more changes later at your comfort. 

This PR has been sitting here for a while, which is not ideal.

Sboonny commented:
I am confused by the ping 🤷. My approval is _still_ there, _you_ are the one who said "this isn't going to get merged without redesign", when I brought this PR at our dev meeting.

There is a C option.

C) Tom is working on a similar feature which put icons next to the map, based on the progress. Create a meeting with Tom and Ahmad and decide what to do with this PR.

I have brought this up every time I could, and it wasn't pushed forward. Maybe without me, it will finally be pushed forward.

----

Side notes to help with moving this forward:

> The tests need to be updated on this PR. The E2E tests need to be fixed (they are not flaky).

True, only currently that the tests aren't flaky. But even with updating the tests, the code needs to update:

- needs to be more dynamic, so it doesn't mind if another certification is added.
- the strings need to be translatable.

---

Note:

I know this isn't "professional", but it will be humane to not bother the people who tried so hard to make this PR happen, either the OP with their patience and working hard to make this happen, or the person who tried what he can to get this merged.

And the person who **blocked** this, should understand how they may feel, and start taking the mantle himself and make this feature happen. 😄

I may be white knighting the OP, and they may want to work on this again, which IDK about 🤷‍♂️, but this is definitely what I feel. 👍

arantebw commented:
I don't mind to continue (resume) working on this PR. If anyone could provide me at least some pointers of what are things that I need to work on to make this PR ready to be merged and passed all of the tests.

ahmadabdolsaheb commented:
Thank you for your patience everyone. I will meet with @Sboonny and @moT01 this week to finalize this pull request.

arantebw commented:
> Hey @arantebw - sorry for the delay on this, and thanks for all the hard work 🎉 It's very well done 👍 I think the PR has sat here for some time because we weren't really sure what we wanted. We labelled the linked issue with the `help wanted` tag before we had a solution figured out - that was a mistake on our part, so our apologies for that 🙏 What we've decided to do is merge this as-is so you get credit for the work - but we are going to revert it because we don't want this UI in production at the moment. What we want looks more like this: #50412. I would ask if you are interested in working on that, but I want to make sure we have enough of the details figured out it before we start the work this time. Feel free to weigh-in over there, you may have some insight. Much of the code from this PR can likely be reused for that as it will use a lot of the same information - so your work here won't go unused.
> 
>   Let us know if you have any thoughts or questions. Thanks for contributing to freeCodeCamp 🎉 and happy coding 👨‍💻

I can totally understand the situation and your decision, I think it's the right thing to do. 👍🏼 

I would love to try and help to implement that feature.

ahmaxed commented:
Thank you for your understanding, here is the related issue for the feature: https://github.com/freeCodeCamp/freeCodeCamp/issues/50412

The requirements are being finalized. Once that is settled, please let us know if you need any help/clarification with the implementation. 

I will close this pr as result. Please feel free to re-open it if you think it has been closed by mistake.

ahmaxed committed:
Merge branch mane into feat/progress-indicator

ahmaxed commented:
I re-opened the pr we could fix it , merge it, give credit to @arantebw, and revert it.
That way the codebase would have the logic if we need it down the line.

ahmaxed committed:
feat: remove superblock title from progress Indicator

ahmaxed commented:
@raisedadead merg it :)
